cpp/examples/arrow/join_example.cc

```diff
@@ -0,0 +1,194 @@
+// intended to be paired with the documentation.
+
+#include <arrow/api.h>
+#include <arrow/compute/api.h>
+#include <arrow/compute/exec/exec_plan.h>
+#include <arrow/compute/exec/expression.h>
+#include <arrow/csv/api.h>
+
+#include <arrow/dataset/dataset.h>
+#include <arrow/dataset/plan.h>
+#include <arrow/dataset/scanner.h>
+
+#include <arrow/io/interfaces.h>
+#include <arrow/io/memory.h>
+#include <arrow/io/stdio.h>
+
+#include <arrow/filesystem/filesystem.h>
+
+#include <arrow/result.h>
+#include <arrow/status.h>
+
+#include <arrow/util/vector.h>
+
+#include <iostream>
+#include <vector>
+
+namespace ds = arrow::dataset;
+namespace cp = arrow::compute;
+
+#define ABORT_ON_FAILURE(expr)                     \
+  do {                                             \
+    arrow::Status status_ = (expr);                \
+    if (!status_.ok()) {                           \
+      std::cerr << status_.message() << std::endl; \
+      abort();                                     \
+    }                                              \
+  } while (0);
+
+std::string GetDataAsCsvString(std::string relation) {
+  std::string data_str = "";
+  if (relation == "l") {
+    data_str = R"csv(lkey,shared,ldistinct
+1,4,7
+2,5,8
+11,20,21
+3,6,9)csv";
+  } else if (relation == "r") {
+    data_str = R"csv(rkey,shared,rdistinct
+1,10,13
+124,10,11
+2,11,14
+3,12,15)csv";
+  } else {
+    return data_str;
+  }
+  return data_str;
+}
+
+arrow::Result<std::shared_ptr<arrow::Table>> GetTableFromExecBatches(
+    const std::shared_ptr<arrow::Schema>& schema,
+    const std::vector<arrow::compute::ExecBatch>& exec_batches) {
+  arrow::RecordBatchVector batches;
+  for (const auto& batch : exec_batches) {
+    ARROW_ASSIGN_OR_RAISE(auto rb, batch.ToRecordBatch(schema));
+    batches.push_back(std::move(rb));
+  }
+  return arrow::Table::FromRecordBatches(schema, batches);
+}
+
+arrow::Result<std::shared_ptr<arrow::dataset::Dataset>> CreateDataSetFromCSVData(
+    std::string relation) {
+  arrow::io::IOContext io_context = arrow::io::default_io_context();
+  std::shared_ptr<arrow::io::InputStream> input;
+  std::string csv_data = GetDataAsCsvString(relation);
+  std::cout << "CSV DATA : " << relation << std::endl;
+  std::cout << csv_data << std::endl;
+  arrow::util::string_view sv = csv_data;
+  input = std::make_shared<arrow::io::BufferReader>(sv);
+  auto read_options = arrow::csv::ReadOptions::Defaults();
+  auto parse_options = arrow::csv::ParseOptions::Defaults();
+  auto convert_options = arrow::csv::ConvertOptions::Defaults();
+
+  // Instantiate TableReader from input stream and options
+  ARROW_ASSIGN_OR_RAISE(std::shared_ptr<arrow::csv::TableReader> table_reader,
+                        arrow::csv::TableReader::Make(io_context, input, read_options,
+                                                      parse_options, convert_options));
+
+  std::shared_ptr<arrow::csv::TableReader> reader = table_reader;
```

Comment:
```suggestion
  const std::shared_ptr<arrow::csv::TableReader>& reader = table_reader;
```

Do you need this alias?  It is only used in one spot below and `table_reader` is a pretty reasonable name already

Reply:
Yes, we can keep it `table_reader`, I agree with you. Probably a negligible naming by me :)